cnf-certification-test/lifecycle/scaling/statefulset_scaling.go

```diff
@@ -0,0 +1,194 @@
+	if statefulset.Spec.Replicas != nil {
+		replicas = *statefulset.Spec.Replicas
+	} else {
+		replicas = 1
+	}
+
+	if replicas <= 1 {
+		// scale up
+		replicas++
+		if !scaleStateFulsetHelper(clients, ssClients, statefulset, replicas, timeout, true) {
+			logrus.Error("can't scale statefulset =", namespace, ":", name)
+			return false
+		}
+		// scale down
+		replicas--
+		if !scaleStateFulsetHelper(clients, ssClients, statefulset, replicas, timeout, false) {
+			logrus.Error("can't scale statefulset =", namespace, ":", name)
+			return false
+		}
+	} else {
+		// scale down
+		replicas--
+		if !scaleStateFulsetHelper(clients, ssClients, statefulset, replicas, timeout, false) {
+			logrus.Error("can't scale statefulset =", namespace, ":", name)
+			return false
+		} // scale up
+		replicas++
+		if !scaleStateFulsetHelper(clients, ssClients, statefulset, replicas, timeout, true) {
+			logrus.Error("can't scale statefulset =", namespace, ":", name)
+			return false
+		}
+	}
+	return true
+}
+
+func scaleStateFulsetHelper(clients *clientsholder.ClientsHolder, ssClient v1.StatefulSetInterface, statefulset *v1app.StatefulSet, replicas int32, timeout time.Duration, up bool) bool {
+	if up {
+		logrus.Trace("scale UP statefulset to ", replicas, " replicas ")
+	} else {
+		logrus.Trace("scale DOWN statefulset to ", replicas, " replicas ")
+	}
+
+	name := statefulset.Name
+	namespace := statefulset.Namespace
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// Retrieve the latest version of statefulset before attempting update
+		// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+		ss, err := ssClient.Get(context.TODO(), name, v1machinery.GetOptions{})
+		if err != nil {
+			logrus.Error("failed to get latest version of statefulset ", namespace, ":", name)
+			return err
+		}
+		ss.Spec.Replicas = &replicas
+		_, err = clients.AppsClients.StatefulSets(namespace).Update(context.TODO(), ss, v1machinery.UpdateOptions{})
+		if err != nil {
+			logrus.Error("can't update statefulset ", namespace, ":", name)
+			return err
+		}
+		if !podsets.WaitForStatefulSetReady(namespace, name, timeout) {
+			logrus.Error("can't update statefulset ", namespace, ":", name)
+			return errors.New("can't update statefulset")
+		}
+		return nil
+	})
+	if retryErr != nil {
+		logrus.Error("can't scale statefulset ", namespace, ":", name, " error=", retryErr)
+		return false
+	}
+	return true
+}
+
+func TestScaleHpaStatefulSet(statefulset *v1app.StatefulSet, hpa *v1autoscaling.HorizontalPodAutoscaler, timeout time.Duration) bool {
+	clients := clientsholder.GetClientsHolder()
+	hpaName := hpa.Name
+	name, namespace := statefulset.Name, statefulset.Namespace
+	hpscaler := clients.K8sClient.AutoscalingV1().HorizontalPodAutoscalers(namespace)
+	var min int32
+	if hpa.Spec.MinReplicas != nil {
+		min = *hpa.Spec.MinReplicas
+	} else {
+		min = 1
+	}
+	max := hpa.Spec.MaxReplicas
+	if min <= 1 {
+		// scale up
+		min++
+		max++
```

Comment:
This implementation seems a bit different than the one in the old repo, where you used the Replica number to set both min and max with it:
```
	replicaCount := podset.Replicas
	podsetscale := *podset
	if podsetscale.Hpa.HpaName != "" {
		podsetscale.Hpa.MinReplicas = replicaCount - 1
		podsetscale.Hpa.MaxReplicas = replicaCount - 1
```
Is this intentional?

Reply:
last repo, we assumed min == max; it's not necessary. We can have different numbers for each one, 
the availability of this statefulset means we need at least min replicas in ready state.

Reply:
I actually thought that, in the old implementation, both min & max were set to the same value as the only way to force the autoscaler to increase/decrease the number of replicas for that deployment/statefulset. Was I wrong @aabughosh ?

So in this new implementation, I don't think it's guaranteed that the actual number of replicas will change if you modify both min and max values, as it will ultimately depend on the configured cpu/mem thresholds for that deployment/statefulset. Am I missing something?